### PR TITLE
Fix #105 issue - missing edges between nodes

### LINKED ISSUE #105: GraphMachine does not include all edges in dot file
I think I found a bug. Code below should produce graph with 4 directional edges, but it gets only one.

```python
from transitions.extensions import GraphMachine as Machine

class AToB(object):
    pass

transitions = [
    ['event_0', 'a', 'b'],
    ['event_1', 'a', 'b'],
    ['event_2', 'a', 'b'],
    ['event_3', 'a', 'b'],
]

machine = Machine(
    AToB(),
    states=['a', 'b'],
    transitions=transitions,
    initial='a',
    auto_transitions=False,
)

machine.graph.draw('/tmp/missing_edges.png', prog='dot')
```

transitions==0.4.0

### PR DESCRIPTION
**Pull request by aisbaa**
Simple fix for missing edges in graph machine. Let me know if different approach is also desired to this problem.

### CODE REVIEW

**Comment by aleneum** [comment]
Unfortunately it is not that easy. The check you want to remove is required for nested graphs where the parent has a valid transition to it's first child and the child itself also has such a transition (for instance because of '*'->parent.child1). Removing it will cause duplicated edges for all *-transitions in nested graphs. There has not been a test for that yet, sorry.

However, I do think the current 'duplicate fix' is not sufficient at all if it prevents having multiple edges with different names. It also 'hides' the edge from parent to child. This is a limitation in graphviz/dot (or how it deals with edges from cluster to nodes) I could not work around yet.

Long story short: I think we should merge the pull request to allow multiple edges and I will open  a new issue to find a better to solution to deal with cluster-to-node-edges.

**Comment by tyarkoni** [comment]
Okay, but in view of the nesting issue, maybe we should just go with my suggestion in #105 to only update the labels rather than allowing multiple edges between the same vertices. Is there a downside to that approach? I'm not sure we need to allow multigraphs in this context. Any FSM with a reasonable number of states is already going to look very messy as is...

**Comment by wtgee** [comment]
I'm with @tyarkoni as I think the labels idea is a little better although that still gets a little crowded on any reasonably sized FSM.

**Comment by aleneum** [comment]
Clutter is a good argument. How should it be annotated then? Like

`event[condition&condition&!unless] || event2[condition&!unless&!unless]`  ?

**Comment by aleneum** [comment]
Example:

```python
states = ['standing', 'walking', {'name': 'caffeinated', 'children':['dithering', 'running']}]
transitions = [
  ['walk', 'standing', 'walking'],
  ['go', 'standing', 'walking'],
  ['stop', 'walking', 'standing'],
  {'trigger': 'drink', 'source': '*', 'dest': 'caffeinated_dithering',
   'conditions':'is_hot', 'unless': 'is_too_hot'},
  ['walk', 'caffeinated_dithering', 'caffeinated_running'],
  ['relax', 'caffeinated', 'standing']
]

machine = GraphMachine(model=Matter(), 
                       states=states, 
                       transitions=transitions, 
                       auto_transitions=False, 
                       initial='standing', 
                       title="Mood Matrix",
                       show_conditions=True)
```
leads to

state

**Comment by wtgee** [comment]
I just tested on my main machine which is modest (I think?) at 11 states. Each state has one `check_safety` condition and some of the states have an additional condition. The generated graph is pretty much too large to use. But I guess if you printed it and put it in a wall for reference it would be handy. :)

I expect not too many people will use this on large machines but it's definitely a feature we should support.

**Comment by wtgee** [comment]
After looking at my large graph closer, either there is a bug in the code or this will be really helpful in finding conditions that are applied multiple times.

If you look at the `park` transition on the top it has the `check_safety` condition listed multiple times and I'm pretty sure my code isn't doing that. I'll look into it more and update shortly.

State File

**Comment by aleneum** [comment]
yeah, I fixed that but wanted to wait for your oppionion about 

> event[condition&condition&!unless] || event2[condition&!unless&!unless]

If you are okay with it I would pull @aisbaa test and commit my changes to make it work with nested graphs

**Comment by wtgee** [comment]
@aleneum Got it, sorry. I'm with it now. Is your fix small that you can post a gist here so I can see run it on my actual machine?

**Comment by aleneum** [comment]
The relevant part is just two lines in extenstion/diagram.py

https://gist.github.com/aleneum/602def83e94bf4ee38907d8151cef9b6

problem is, that `label` is falsely reused.

[this](https://gist.github.com/aleneum/602def83e94bf4ee38907d8151cef9b6#file-fix105-diff-L10) and [this](https://gist.github.com/aleneum/602def83e94bf4ee38907d8151cef9b6#file-fix105-diff-L29) should be enough.

**Comment by wtgee** [comment]
Great, thanks. Looks a lot better for sure. :+1:

On Wed, Jun 1, 2016 at 6:48 AM, Alexander Neumann <notifications@github.com>
wrote:

> The relevant part is just two lines in extenstion/diagram.py
>
> https://gist.github.com/aleneum/602def83e94bf4ee38907d8151cef9b6
>
> problem is, that label is falsely reused.
>
> this
> <https://gist.github.com/aleneum/602def83e94bf4ee38907d8151cef9b6#file-fix105-diff-L10>
> and this
> <https://gist.github.com/aleneum/602def83e94bf4ee38907d8151cef9b6#file-fix105-diff-L29>
> should be enough.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tyarkoni/transitions/pull/107#issuecomment-223054548>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAEUUI5DZATKO9RKEVUvKiTfJfSPI7Moks5qHbfygaJpZM4IqgYc>
> .
>



-- 
~Wilfred Tyler Gee
